### PR TITLE
Add ES2015 class serialization support

### PR DESCRIPTION
**Pull request by trueadm**
Release notes: Adds ES2015 class support

This PR adds ES2015 class serialization support to Prepack. I've added tests and run many local experiments to ensure the following are supported:

- Constructors (and empty constructors)
- Inheritance
- Super
- Getters
- Setters
- String method names
- Symbol method names
- `new Class()`
- Additional functions

### CODE REVIEW

File: src/serializer/ResidualFunctions.js [hunk 1]
```diff
@@ -35,10 +37,11 @@ import { nullExpression } from "../utils/internalizer.js";
 import type { LocationService } from "./types.js";
 import { Referentializer } from "./Referentializer.js";
 import { getOrDefault } from "./utils.js";
+import ObjectValue from "../values/ObjectValue";
```
[the remark this answers is not in the document]
Comment (trueadm) [reply src/serializer/ResidualFunctions.js]: I didn't actually explicitly write this, I think it was VSCode auto importing :/

File: src/serializer/ResidualFunctions.js [hunk 2]
```diff
@@ -292,15 +297,31 @@ export class ResidualFunctions {
...
-        strictFunctionBodies.push(funcNode);
+      let funcOrClassNode;
+      let { isClassMethod, classSuper } = instance;
+
+      if (isClassMethod) {
+        let homeObject = functionValue.$HomeObject;
+        invariant(homeObject instanceof ObjectValue);
+        invariant(
+          functionValue.$FunctionKind !== "classConstructor",
+          "Class constructors cannot be additional functions"
```
[the remark this answers is not in the document]
Comment (trueadm) [reply src/serializer/ResidualFunctions.js]: That's a good point. I'll make this a diagnostic error rather than invariant.

File: src/serializer/ResidualFunctions.js [hunk 3]
```diff
@@ -292,15 +297,31 @@ export class ResidualFunctions {
...
+        );
+        let methodName = Get(this.realm, functionValue, "name");
+        invariant(methodName instanceof StringValue);
```
[the remark this answers is not in the document]
Comment (trueadm) [reply src/serializer/ResidualFunctions.js]: Ah yes, I forgot about computed class method names. What would be a better way of dealing with them?

File: src/serializer/ResidualFunctions.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (trueadm) [reply src/serializer/ResidualFunctions.js]: Actually, I think we can support constructors too. This PR is still a WIP, so I'll add it to my TODO list.

File: src/serializer/ResidualFunctions.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (hermanventer) [reply src/serializer/ResidualFunctions.js]: Ooh, salivate.

File: src/serializer/ResidualFunctions.js [hunk 4]
```diff
@@ -325,42 +355,102 @@ export class ResidualFunctions {
...
+          let { functionValue, residualFunctionBindings, scopeInstances, classProperties } = instance;
+          let funcOrClassNode;
+
+          if (classProperties !== undefined) {
+            let { classSuperNode, classMethodKeyNode, methodType, classMethodComputed } = classProperties;
+
+            let isConstructor = methodType === "constructor";
+            let homeObject = functionValue.$HomeObject;
+            invariant(homeObject instanceof ObjectValue);
+            invariant(classMethodKeyNode && (t.isExpression(classMethodKeyNode) || t.isIdentifier(classMethodKeyNode)));
```
Comment (cblappert) [inline src/serializer/ResidualFunctions.js]: prefer separate invariants to conjunctions

File: src/serializer/ResidualHeapSerializer.js [hunk 5]
```diff
@@ -1117,9 +1126,106 @@ export class ResidualHeapSerializer {
...
+        // we visit all of these as these, as they form part of the constructor for the class
+        // but we want to serialize them out as class methods in the ResidualFunctions
+        // rather than via the typical Prepack route of serializing properties
+        for (let [key] of val.properties) {
+          if (!this.residualHeapInspector.canIgnoreProperty(val, key)) {
+            let propertyValue = Get(this.realm, val, key);
+
+            if (key === "arguments" || key === "length" || key === "name" || key === "caller" || key === "prototype") {
+              properties.delete(key);
+              this.serializedValues.add(propertyValue);
```
Comment (cblappert) [inline src/serializer/ResidualHeapSerializer.js]: Prefer changing the visitor to understand classes to explicitly adding to serializedValues. (make a new visitValueClass method that visitValue or visitValueFunction calls to when it detects a class that decides not to visit the arguments, length, name etc).

File: src/serializer/ResidualHeapSerializer.js [hunk 6]
```diff
@@ -1117,9 +1126,106 @@ export class ResidualHeapSerializer {
...
+          }
+        }
 
-    undelay();
+        let handleClassMethodValue = (classMethod, methodProps: MethodProperties) => {
+          invariant(classMethod instanceof FunctionValue);
+          // skip processing if its the constructor
+          if (classMethod !== val) {
+            this.serializedValues.add(classMethod);
+            this._serializeValueFunction(classMethod, methodProps);
```
Comment (cblappert) [inline src/serializer/ResidualHeapSerializer.js]: @NTillmann, do you think there's a better way to do this/can you think of any problems with not going through serializeValue?

File: src/serializer/ResidualHeapSerializer.js (hunk 5 above)
Comment (trueadm) [reply src/serializer/ResidualHeapSerializer.js]: I can make this change, but I remember running into issues when I didn't visit `arguments` and `caller` (from memory).

Comment (trueadm) [reply src/serializer/ResidualHeapSerializer.js]: @cblappert I don't think this is as simple as not visiting it. 

`_emitObjectProperties()` goes through all the properties on the class object and all its methods (as expected), including `arguments`, `length', `key' etc, regardless if they've been visited or not – and tries to serialize them all. I don't want `_emitObjectProperties` to serialize these values – because we create the AST for them with entirely separate logic.

File: src/values/ECMAScriptSourceFunctionValue.js [hunk 7]
```diff
@@ -23,6 +23,8 @@ export default class ECMAScriptSourceFunctionValue extends ECMAScriptFunctionVal
   $Strict: boolean;
   $FormalParameters: Array<BabelNodeLVal>;
   $ECMAScriptCode: BabelNodeBlockStatement;
+  $HasComputedName: boolean;
```
Comment (NTillmann) [inline src/values/ECMAScriptSourceFunctionValue.js]: I am not convinced that `$HasComputedName` and `$HasEmptyConstructor` are set along all paths where an instance of this class is created. Thus, the type should probably be `?: boolean`.

File: src/serializer/ResidualHeapSerializer.js (hunk 6 above)
Comment (NTillmann) [reply src/serializer/ResidualHeapSerializer.js]: Can there be any other reference to the `classMethod`? If so, then that could cause problems.

Comment (NTillmann) [reply src/serializer/ResidualHeapSerializer.js]: We should try to find a test case for that.

File: src/serializer/ResidualHeapSerializer.js [hunk 8]
```diff
@@ -1117,9 +1126,110 @@ export class ResidualHeapSerializer {
...
+              properties.delete(key);
+              this.serializedValues.add(propertyValue);
+            }
+          }
+        }
 
-    undelay();
+        let handleClassMethodValue = (classMethod, methodProps: MethodProperties) => {
+          invariant(classMethod instanceof FunctionValue);
+          // skip processing if its the constructor
```
Comment (NTillmann) [inline src/serializer/ResidualHeapSerializer.js]: it's

File: src/serializer/ResidualHeapSerializer.js (hunk 6 above)
Comment (NTillmann) [reply src/serializer/ResidualHeapSerializer.js]: In serializeValue, you could log a nice user-facing error if such a `classMethod` ever gets there.

File: src/serializer/ResidualHeapVisitor.js [hunk 9]
```diff
@@ -383,8 +394,35 @@ export class ResidualHeapVisitor {
         if (functionInfo.modified.has(innerName)) residualFunctionBinding.modified = true;
       }
     });
-
+    let useClassProperties = false;
+    let isConstructor = false;
+    let isComputed = val.$HasComputedName;
+
+    if (val.$HomeObject instanceof AbstractValue) {
```
Comment (NTillmann) [inline src/serializer/ResidualHeapVisitor.js]: The Flow-type of `$HomeObject` doesn't allow for that. It's either `undefined`, or an `ObjectValue`. You should assert that here, and simplify the control-flow.

File: src/serializer/ResidualHeapVisitor.js [hunk 10]
```diff
@@ -383,8 +394,33 @@ export class ResidualHeapVisitor {
...
+        if (constructorValue.$FunctionKind === "classConstructor") {
+          useClassProperties = true;
+          if (val.$FunctionKind === "classConstructor") {
+            isConstructor = true;
+          }
+        }
+      }
+    }
     this.functionInstances.set(val, {
+      classProperties: useClassProperties
```
Comment (sebmarkbage) [inline src/serializer/ResidualHeapVisitor.js]: The name suggests to me that it would contain a list of properties. Maybe call it something like `methodData` or something?

File: src/serializer/ResidualFunctions.js [hunk 11]
```diff
@@ -288,19 +291,46 @@ export class ResidualFunctions {
...
+        invariant(classMethodKeyNode && (t.isExpression(classMethodKeyNode) || t.isIdentifier(classMethodKeyNode)));
+        // we use the $HomeObject as the key to get the class expression ast node
+        funcOrClassNode = this._getOrCreateClassNode(homeObject);
+        let classMethod = t.classMethod(methodType, classMethodKeyNode, funcParams, functionBody, classMethodComputed);
+        // add the class method to the class expression node body
+        if (isConstructor) {
+          funcOrClassNode.body.body.unshift(classMethod);
+        } else {
+          funcOrClassNode.body.body.push(classMethod);
+        }
```
Comment (sebmarkbage) [inline src/serializer/ResidualFunctions.js]: Just because this method was originally defined on this class, doesn't mean it still is the same one.

E.g.

```js
class Foo {
  bar() {
  }
  baz() {
  }
}
Foo.prototype.bar = Foo.prototype.bar.bind({});
var baz = Foo.prototype.baz;
delete Foo.prototype.baz;
result = { Foo, baz };
```

We should test that the property on the class is still the correct one.

It's probably fine not supporting this yet, but we should at least fatal and not generate subtly wrong code.

Comment (sebmarkbage) [reply src/serializer/ResidualFunctions.js]: Static class methods also have home objects and will end up here but they will be serialized as an instance method instead which is wrong.

File: src/evaluators/ClassDeclaration.js [hunk 12]
```diff
@@ -169,6 +170,7 @@ export function ClassDefinitionEvaluation(
 
   // 10. If constructor is empty, then,
   if (constructor instanceof EmptyValue) {
+    emptyConstructor = true;
```
[the remark this answers is not in the document]
Comment (trueadm) [reply src/evaluators/ClassDeclaration.js]: I tried that, it was too messy to read the AST and infer this.

File: src/serializer/ResidualFunctions.js (hunk 11 above)
Comment (trueadm) [reply src/serializer/ResidualFunctions.js]: Static methods are supported fine – they were tested. They are actually properties of the constructor function.

File: src/methods/properties.js [hunk 13]
```diff
@@ -1274,6 +1274,9 @@ export class PropertiesImplementation {
       // 3. Perform SetFunctionName(methodDef.[[closure]], methodDef.[[key]]).
       Functions.SetFunctionName(realm, methodDef.$Closure, methodDef.$Key);
 
+      // If the AST name was computed, give the hint to the closure
+      methodDef.$Closure.$HasComputedName = !!MethodDefinition.computed;
```
[the remark this answers is not in the document]
Comment (trueadm) [reply src/methods/properties.js]: I'm not sure I follow here – how is this related to computed method names? Your above example seems to work fine too?

File: src/serializer/ResidualFunctions.js (hunk 11 above)
Comment (sebmarkbage) [reply src/serializer/ResidualFunctions.js]: If so, can you add a unit test for it? With and without a super call in the static method.

Comment (trueadm) [reply src/serializer/ResidualFunctions.js]: @sebmarkbage Sure. Although I'm not sure what is special about the `super` call. Prepack doesn't touch it as we don't ever evaluate it.

File: src/serializer/ResidualHeapVisitor.js [hunk 14]
```diff
@@ -136,6 +136,17 @@ export class ResidualHeapVisitor {
...
+      // we don't want to visit these as we handle the serialization ourselves
+      // via a different logic route for classes
+      if (
+        obj.$FunctionKind === "classConstructor" &&
+        (propertyBindingKey === "arguments" ||
+          propertyBindingKey === "length" ||
+          propertyBindingKey === "name" ||
+          propertyBindingKey === "caller")
+      ) {
+        continue;
```
[the remark this answers is not in the document]
Comment (trueadm) [reply src/serializer/ResidualHeapVisitor.js]: I'll add tests for these cases and try and fix them accordingly.

File: src/methods/properties.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (sebmarkbage) [reply src/methods/properties.js]: Basically, I'm saying that it doesn't matter how it was originally created. It matters more what is currently there.

So `class Foo { ['foo']() { } }` should be able to serialize as `class Foo { foo() { } }`.

I'm not sure how I understand how later mutations to the prototype gets serialized. Like this:

```js
Foo.prototype.bar = function() { };
```

**Comment by trueadm** [comment]
@sebmarkbage Even better, if you can come up with more good short example tests, please stick them on here and I'll build tests around them and fix them accordingly. :)

File: src/methods/properties.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (trueadm) [reply src/methods/properties.js]: Both cases are handled as you put above. The computed method gets turned into a string property and the later added `bar` also gets added. :) There are tests for your first case already, I'll add some for the second case tomorrow.

File: src/serializer/ResidualHeapSerializer.js [hunk 15]
```diff
@@ -1279,9 +1420,30 @@ export class ResidualHeapSerializer {
           proto !== this.realm.intrinsics.ObjectPrototype &&
           this._findLastObjectPrototype(val) === this.realm.intrinsics.ObjectPrototype &&
           proto instanceof ObjectValue;
+        let isClass = false;
+        let classPrototype;
+        // if the object has a prototype that was a class (by check the constructor on the prototype)
+        if (proto instanceof ObjectValue && proto !== this.realm.intrinsics.ObjectPrototype) {
+          // check if this object is actually based from a class (by checking the constructor)
+          let valConstructor = Get(this.realm, val, "constructor");
```
Comment (sebmarkbage) [inline src/serializer/ResidualHeapSerializer.js]: Not every instance of a class has a constructor. E.g.

```js
class Foo { }
delete Foo.prototype.constructor;
```

This could also be a getter which could be problematic.

The prototype on the constructor also not be the same anymore. E.g.

```js
class Foo {}
result = new Foo();
Foo.prototype = {};
```

or

```js
class Foo { };
result = Object.create({ constructor: Foo });
```

The only reason this needs to be serialized as a class is if any function remains that has this as the HomeObject. Otherwise it can just be a normal object.

Another solution might be to just tag the `ObjectValue` of the prototype at construction time just like the constructor is.

File: src/serializer/ResidualHeapVisitor.js [hunk 16]
```diff
@@ -383,8 +394,33 @@ export class ResidualHeapVisitor {
...
       }
     });
-
+    let useClassProperties = false;
+    let isConstructor = false;
+    let isComputed = !!val.$HasComputedName;
+
+    if (val.$HomeObject instanceof ObjectValue) {
+      // determine if our $HomeObject is actually a ES2015 class by looking at constructor
+      let constructorValue = Get(this.realm, val.$HomeObject, "constructor");
```
Comment (sebmarkbage) [inline src/serializer/ResidualHeapVisitor.js]: Same concern here as the other one. This is not a "safe" property to check.

File: src/serializer/ResidualFunctions.js (hunk 11 above)
Comment (sebmarkbage) [reply src/serializer/ResidualFunctions.js]: Super is special because that can't be serialized as:

```js
class Foo extends Bar {}
Foo.bar = function() { super(); };
```

File: src/serializer/ResidualHeapSerializer.js (hunk 15 above)
Comment (trueadm) [reply src/serializer/ResidualHeapSerializer.js]: I'll address this and the other concerns you had in a follow up tomorrow along with some tests. Thanks :)

**Comment by trueadm** [comment]
@yinghuitan Definitely agree that the code should be in its own module. I think we can probably do that as a follow up PR though, if that's okay.

**Comment by yinghuitan** [comment]
@trueadm, I am fine with doing it in later PR.

**Comment by NTillmann** [comment]
This pull request has been around for a while. What is left to do?

**Comment by trueadm** [comment]
Would be awesome if @sebmarkbage could give this a once over again. :)

**Comment by trueadm** [comment]
@NTillmann @sebmarkbage Can we merge this PR if everyone is happy, it's been lying around for some time now and would be useful if instead any issues we found could be added as follow ups instead of blocking this PR?

File: src/evaluators/ClassDeclaration.js [hunk 17]
```diff
@@ -150,11 +150,15 @@ export function ClassDefinitionEvaluation(
     // 7. Let proto be ObjectCreate(protoParent).
     let proto = Create.ObjectCreate(realm, protoParent);
 
+    // Provide a hint that this prototype is that of a class
```
Comment (hermanventer) [inline src/evaluators/ClassDeclaration.js]: Why only a hint?

File: src/serializer/ResidualFunctions.js (hunk 11 above)
Comment (trueadm) [reply src/serializer/ResidualFunctions.js]: This is fixed now.

File: src/methods/properties.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (trueadm) [reply src/methods/properties.js]: XmeineXAlliance:

File: src/serializer/ResidualFunctions.js (hunk 11 above)
Comment (hermanventer) [reply src/serializer/ResidualFunctions.js]: Is there a test case for this?

File: src/serializer/ResidualFunctions.js [hunk 18]
```diff
@@ -293,18 +296,32 @@ export class ResidualFunctions {
...
 
         let isConstructor = methodType === "constructor";
-        let homeObject = functionValue.$HomeObject;
-        invariant(homeObject instanceof ObjectValue);
+        invariant(classPrototype instanceof ObjectValue);
         invariant(classMethodKeyNode && (t.isExpression(classMethodKeyNode) || t.isIdentifier(classMethodKeyNode)));
-        // we use the $HomeObject as the key to get the class expression ast node
-        funcOrClassNode = this._getOrCreateClassNode(homeObject);
-        let classMethod = t.classMethod(methodType, classMethodKeyNode, funcParams, functionBody, classMethodComputed);
+        // we use the classPrototype as the key to get the class expression ast node
```
Comment (hermanventer) [inline src/serializer/ResidualFunctions.js]: Are there checks to ensure that the classPrototype uniquely corresponds to a single class expression AST node?

File: src/serializer/ResidualHeapVisitor.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (trueadm) [reply src/serializer/ResidualHeapVisitor.js]: So for your first case: `static bar = cycle;` is invalid syntax currently, so skipping this.
Second case: this seems handled currently. Test for this needs adding.
Third case: I never got around to this, so this needs to be done and tests need adding.

**Comment by trueadm** [comment]
@hermanventer I've not strictly gone through this PR according to the spec. I instead added functionality and then added instances of what the spec states to match the functionality via manual testing. I feel this PR needs a bit more work in terms of additional tests and a few bits of missing functionality relating to when the class has been constructed, i.e. `var a = new ClassThing(); a.doSomething()`.

**Comment by NTillmann** [comment]
@sebmarkbage, anything we can do to help move this forward?

File: src/serializer/ResidualHeapSerializer.js [hunk 19]
```diff
@@ -699,6 +714,24 @@ export class ResidualHeapSerializer {
...
+      // so we need to log a nice error message to the user
+      if (classMethodInstance.methodType !== "constructor") {
+        let error = new CompilerDiagnostic(
+          "a class method incorrectly went through the serializeValue() code path",
+          val.$ECMAScriptCode.loc,
+          "PP0021",
+          "FatalError"
+        );
+        this.realm.handleError(error);
+        throw new FatalError();
```
Comment (sebmarkbage) [inline src/serializer/ResidualHeapSerializer.js]: Can this happen? If it can never happen, then this should be an `invariant`, but if it can, then we should have a test for it.

File: src/evaluators/ClassDeclaration.js (hunk 17 above)
Comment (sebmarkbage) [reply src/evaluators/ClassDeclaration.js]: It is kind of a hint because it is a hint that we should try to serialize this as a class body instead of an object literal. It is possible that this doesn't need to be serialized as a class and that this is unnecessary (e.g. if it has no `super` calls). It is also possible that some functions that are not defined as classes in source would be better off being serialized as a class. E.g. a lowered Babel class can probably be turned back into a class.

We could determine either one based on some other heuristic but for now it is a hint.

File: src/serializer/utils.js [hunk 20]
```diff
@@ -81,3 +82,24 @@ export function getOrDefault<K, V>(map: Map<K, V>, key: K, defaultFn: () => V):
...
+        func(propertyNameOrSymbol, descriptor.get, "get");
+      }
+      if (descriptor.set !== undefined) {
+        func(propertyNameOrSymbol, descriptor.set, "set");
+      }
+    }
+  }
+}
+
+export const ClassProprtiesToIgnore: Set<string> = new Set(["arguments", "name", "caller", "length"]);
```
[the remark this answers is not in the document]
Comment (sebmarkbage) [reply src/serializer/utils.js]: It is not always safe to ignore these. E.g. this example:

```js
class Foo {
  
}

Object.defineProperty(
  Foo,
  'length',
  { value: 2}
);
```

If you look at how this is handled for other functions, you see that they're only conditionally ignored.

https://github.com/facebook/prepack/blob/ed40a493e2dff76e61158458e553ad6e7ce975bf/src/serializer/ResidualHeapInspector.js#L82-L106

File: src/serializer/utils.js [hunk 21]
```diff
@@ -81,3 +82,24 @@ export function getOrDefault<K, V>(map: Map<K, V>, key: K, defaultFn: () => V):
   invariant(value !== undefined);
   return value;
 }
+
+export function withDescriptorValue(
+  propertyNameOrSymbol: string | SymbolValue,
+  descriptor: void | Descriptor,
+  func: Function
```
Comment (sebmarkbage) [inline src/serializer/utils.js]: This function should be typed using generics and a specific function signature. All these functions passed here are inferred as `any` right now.

File: src/serializer/ResidualHeapSerializer.js [hunk 22]
```diff
@@ -1091,10 +1124,128 @@ export class ResidualHeapSerializer {
...
+  }
+
+  _serializeClassMethod(key: string | SymbolValue, methodFunc: ECMAScriptSourceFunctionValue): void {
+    let classMethodInstance = this.residualClassMethodInstances.get(methodFunc);
+
+    invariant(classMethodInstance !== undefined);
+    if (typeof key === "string") {
+      classMethodInstance.classMethodKeyNode = t.identifier(key);
+      // as we know the method name is a string again, we can remove the computed status
+      classMethodInstance.classMethodComputed = false;
```
Comment (sebmarkbage) [inline src/serializer/ResidualHeapSerializer.js]: How do we know this? If this is a string but an invalid identifier like `class Foo { ["foo bar"]() { } }` then it still must be computed.

Or if I rename this:

```js
class Foo {
  bar() {
  }
}
Foo.prototype['foo bar'] = Foo.prototype.bar;
delete Foo.prototype.bar;
```

File: src/serializer/utils.js (hunk 21 above)
Comment (trueadm) [reply src/serializer/utils.js]: This was done intentionally. Flow just gets confused and outputs about 20 errors if you try and make this something other than `any`. I spoke to the Flow team and they said that Flow struggles with this sort of pattern and its on their roadmap to fix. I'm not sure it's of any value to hack around Flow to make this work right now.

**Comment by trueadm** [comment]
This PR has gotten too unwieldy to properly manage now – it's hard to track what needs to be fixed where because minor changes to a file mark comments as out of date (a terrible auto-GitHub feature in my opinion, it should be manually controlled). So, instead I'll merge this PR and create a new issue with the outstanding issues on them as per comments in this PR. The next PR I setup to fix the issues will also include more extensive tests. This shouldn't block anything as we aren't shipping any code that includes ES2015 classes and the bugs in this are more edge-cases rather than common usage/behaviour.

**Comment by trueadm** [comment]
Follow up issue created: https://github.com/facebook/prepack/issues/1317
